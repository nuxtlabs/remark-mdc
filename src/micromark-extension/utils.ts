import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
