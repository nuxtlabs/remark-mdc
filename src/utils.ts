export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
