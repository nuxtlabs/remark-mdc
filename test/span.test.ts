import { describe, expect } from 'vitest'
import { runMarkdownTests } from './utils'

describe('span', () => {
  runMarkdownTests({
    simple: {
      markdown: '[span]'
    },
    'inside-link': {
      markdown: '[[span]](#href)'
    },
    'with-attributes': {
      markdown: '[span]{#theid.aclass foo="bar" class="anotherclass" class="andother"}',
      expected: '[span]{#theid .aclass.anotherclass.andother foo="bar"}',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children[0].attributes).toEqual({
          id: 'theid',
          class: 'aclass anotherclass andother',
          foo: 'bar'
        })
      }
    },
    '[x]': {
      markdown: '[x]',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children[0].type).toEqual('textComponent')
        expect(ast.children[0].children[0].children[0].value).toEqual('x')
      }
    },
    '[ ]': {
      markdown: '[ ]',
      expected: '[]',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children[0].type).toEqual('textComponent')
        expect(ast.children[0].children[0].children.length).toEqual(0)
      }
    },
    'respect-gfm-check-list': {
      markdown: '- [ ] task 1\n- [X] task 2',
      expected: '-   [ ] task 1\n-   [x] task 2'
    }
  })
})
