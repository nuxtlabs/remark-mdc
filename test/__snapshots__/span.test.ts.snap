// Vitest Snapshot v1

exports[`span > [ ] 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [],
          "data": {
            "hName": "span",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "span",
          "position": {
            "end": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
      ],
      "position": {
        "end": {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`span > [x] 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "position": {
                "end": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "x",
            },
          ],
          "data": {
            "hName": "span",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "span",
          "position": {
            "end": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
      ],
      "position": {
        "end": {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`span > inside-link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "attributes": {},
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 7,
                      "line": 1,
                      "offset": 6,
                    },
                    "start": {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "text",
                  "value": "span",
                },
              ],
              "data": {
                "hName": "span",
                "hProperties": {},
              },
              "fmAttributes": {},
              "name": "span",
              "position": {
                "end": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "textComponent",
            },
          ],
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "#href",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`span > respect-gfm-check-list 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": false,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 13,
                      "line": 1,
                      "offset": 12,
                    },
                    "start": {
                      "column": 7,
                      "line": 1,
                      "offset": 6,
                    },
                  },
                  "type": "text",
                  "value": "task 1",
                },
              ],
              "position": {
                "end": {
                  "column": 13,
                  "line": 1,
                  "offset": 12,
                },
                "start": {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": true,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 13,
                      "line": 2,
                      "offset": 25,
                    },
                    "start": {
                      "column": 7,
                      "line": 2,
                      "offset": 19,
                    },
                  },
                  "type": "text",
                  "value": "task 2",
                },
              ],
              "position": {
                "end": {
                  "column": 13,
                  "line": 2,
                  "offset": 25,
                },
                "start": {
                  "column": 7,
                  "line": 2,
                  "offset": 19,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 13,
              "line": 2,
              "offset": 25,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 13,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 13,
          "line": 2,
          "offset": 25,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 13,
      "line": 2,
      "offset": 25,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`span > simple 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "span",
            },
          ],
          "data": {
            "hName": "span",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "span",
          "position": {
            "end": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
      ],
      "position": {
        "end": {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`span > with-attributes 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "aclass anotherclass andother",
            "foo": "bar",
            "id": "theid",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "span",
            },
          ],
          "data": {
            "hName": "span",
            "hProperties": {
              "class": "aclass anotherclass andother",
              "foo": "bar",
              "id": "theid",
            },
          },
          "fmAttributes": {},
          "name": "span",
          "position": {
            "end": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
      ],
      "position": {
        "end": {
          "column": 70,
          "line": 1,
          "offset": 69,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 70,
      "line": 1,
      "offset": 69,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
